### PR TITLE
[Dashboard] feat: add new team filter component

### LINKED ISSUE #2849: [Dashboard] Update colony teams navigation
### Story

* As a USER, I want to be able to see a high level breakdown of the team structure of the colony.

### Description

Given the focus on financial actions and tools, we want to simplify the switch between teams by introducing a tabbed navigation.

If the number of teams is higher than available width, the kebab option is shown allowing the user to select the full list of top level teams to filter by - max 10 teams should be displayed in the kebab menu, introducing a scrollbar if there are more.
The tabbed navigation includes all top-level teams (first level of teams available in the colony), sorted by reputation in the colony. If there is no reputation, they need to be sorted alphabetically.
Also, we want to provide support for nested teams, including a caret icon which triggers a dropdown navigation if a team has nested teams. Clicking on one option should update the teams shown in the tabbed navigation.

Image

The tabbed navigation always includes an 'All teams' tab. Clicking on it resets the team filter.

The tabbed navigation always includes a '+' tab that opens the side action panel with the Create team action type pre-selected

## Implementation
We need to use path parameter

### This issue is complete when...

- [ ] Dashboard sub header includes team navigation
- [ ] Dashboard sub header includes All teams tab
- [ ] Dashboard sub header includes the Add new team button
- [ ] Selecting one team navigation tab activates domain/team filtering
- [ ] Selecting one nested team updated the displayed teams and activates the domain/team filtering

[Figma team filtering tabs component](https://www.figma.com/design/gac6xVgGNCAge5Bia933dp/Dashboard?node-id=4878-89693&t=SNW5sWiyJk5omQMX-0)
[Figma link](https://www.figma.com/design/gac6xVgGNCAge5Bia933dp/Dashboard?node-id=4859-36367&t=3YmH0OCdhMXflaYC-0)
[Figma link empty state](https://www.figma.com/design/gac6xVgGNCAge5Bia933dp/Dashboard?node-id=4867-40639&t=3YmH0OCdhMXflaYC-0)

**Feature branch**: `feat/enhanced-dashboard`

### PR DESCRIPTION
**Pull request by bassgeta**
## Description

Changing the solution from the dropdown to use the newly designed tabs.
Removing the current dropdown should be done as part of #2860.

[Figma link](https://www.figma.com/design/gac6xVgGNCAge5Bia933dp/Dashboard?node-id=4887-20952&t=XheoZNvOyt62DJXq-4)
image

## Testing

Needs to be tested functionality wise and UI wise.

### Functionality
1. Make sure that when you click a team, it shows up in the search params and everything starts getting filtered
image
2. Clicking on another one should immediately change the filter:
image
3. Clicking on the same one should disable it and highlight `All teams`:
image
4. Clicking on `All teams` when a domain is selected, should disable the filter
5. Clicking on the `+` button should open up the sidebar with the "Create new team" action
image

### Mobile UI
1. Set the screen to a mobile size and verify that you see the arrows:
image
2. Clicking on the right arrow should scroll a bit forward and the left arrow should appear
image
3. Verify that you can click on the right button until you get to the end and see the new team button
image
4. Verify that you can also scroll left with the left button
5. Additionally, you should be able to drag this on mobile

### Desktop UI
1. Open up the dev tools and start resizing! When the teams would take up too much space, a dropdown should appear, displaying the overflowing teams inside it (NOTE: if you go to less than 850px, the mobile view will appear)
image
2. On hover the text should be bold
image
3. When you select a team from the dropdown, it should display on the last spot, pushing the previously last domain into the dropdown
image
4. Choose `azizazizazizazizaziz` and try to expand and shrink the viewport, you should be able to see how items pop out and into the dropdown

Then just try to break it :D 

## Diffs

**New stuff** ✨

* Fully composable `ArrowScroller` component
* New `TeamFilter` component which has mobile and desktop variants separated
* `DesktopTeamFilter` which has the magic disappearing component with the dropdown
* `MobileTeamFilter` which uses the `ArrowScroller`

Resolves  #2849

### CODE REVIEW

**Review by iamsamgibbs — CHANGES_REQUESTED** [review]
Amazing work on this complicated component! It looks and works really great!

I've left one very minor comment regarding a border - which I think is the only "must fix" issue before approving. Everything else in this comment are more edge case / nice to have.

The items in the dropdown go bold on hover which is as expected, but it causes the width of the container to increase. On the figma design the container is the same size when an item is hover and not hovered, so I don't think this is desired. Although I imagine it will also be a bit of a pain to fix this.

https://github.com/user-attachments/assets/d979b4d4-81c9-432f-be17-4c1ebc6e7e4b

I found this weird edge case when the team names will fill it on a first load, but once you resize below the full screen width then they won't go back to filling up the entire space. (Probably enough of an edge case that we don't need to worry about it).

https://github.com/user-attachments/assets/859faeae-9069-4f5c-9fe3-35c76a8dfaa1

I tested adding new teams until it exceeded the width of the page without issue. 🎉

This is more of a design thing I guess, but I would expect mobile to behave it a similar way to desktop with the "active" team showing earlier in the menu on a page load. For example, navigating directly to `http://localhost:9091/planex?team=4` will have the `Rocinante` team selected, but it requires a scroll before it shows.

Screenshot 2024-09-13 at 15 58 17

Screenshot 2024-09-13 at 15 58 30

However, I imagine this would be a huge pain to implement and would require further discussion with product / design.

**Comment by bassgeta** [comment]
> Amazing work on this complicated component! It looks and works really great!
> 
> I've left one very minor comment regarding a border - which I think is the only "must fix" issue before approving. Everything else in this comment are more edge case / nice to have.
> 
> The items in the dropdown go bold on hover which is as expected, but it causes the width of the container to increase. On the figma design the container is the same size when an item is hover and not hovered, so I don't think this is desired. Although I imagine it will also be a bit of a pain to fix this.
> Screen.Recording.2024-09-13.at.16.00.04.mov
> 
> I found this weird edge case when the team names will fill it on a first load, but once you resize below the full screen width then they won't go back to filling up the entire space. (Probably enough of an edge case that we don't need to worry about it).
> Screen.Recording.2024-09-13.at.16.06.57.mov
> 
> I tested adding new teams until it exceeded the width of the page without issue. 🎉
> 
> This is more of a design thing I guess, but I would expect mobile to behave it a similar way to desktop with the "active" team showing earlier in the menu on a page load. For example, navigating directly to `http://localhost:9091/planex?team=4` will have the `Rocinante` team selected, but it requires a scroll before it shows.
> Screenshot 2024-09-13 at 15 58 17 Screenshot 2024-09-13 at 15 58 30
> 
> However, I imagine this would be a huge pain to implement and would require further discussion with product / design.

Yeah it does behave a bit weirdly when you resize the window back and forth, which would probably be a bit of a pain to fix, I was thinking of rerendering it whenever `window.width` changes and see if that would work. So I'll give it a shot if maybe that would solve it :D 

Regarding the last point for the mobile scroller, I think it's a valid point, but maybe we can do that separately since it'll involve at least **some** extra shennanigans.

File: src/components/v5/shared/TeamFilter/partials/AllTeamsItem.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  const handleClick = () => {
+    searchParams.delete(TEAM_SEARCH_PARAM);
+    setSearchParams(searchParams);
+  };
+
+  return (
+    <button
+      type="button"
+      className={clsx(
+        'border-r border-solid border-gray-200 px-4 py-2 text-3',
```
Comment (mmioana) [inline src/components/v5/shared/TeamFilter/partials/AllTeamsItem.tsx:34]: By using `text-3` the `font-semibold` or `font-medium` get overridden
So either use `!font-semibold` and `!font-medium` OR replace `text-3` with `text-sm`

File: src/components/v5/shared/TeamFilter/partials/CreateNewTeamItem.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    });
+  };
+
+  return (
+    <button
+      type="button"
+      className="border-gray-200 bg-base-white px-4 py-2 font-medium text-gray-400 text-3 hover:bg-gray-50 hover:text-gray-900"
+      onClick={handleClick}
+    >
+      <Plus size={16} />
```
Comment (mmioana) [inline src/components/v5/shared/TeamFilter/partials/CreateNewTeamItem.tsx:27]: According to Figma this should be `18`

File: src/components/v5/shared/TeamFilter/partials/DesktopTeamFilter/DesktopTeamFilter.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    return getOrderedDomains({
+      maxDomains: overflowingItemIndex,
+      domains: allDomains,
+      selectedDomain,
+    });
+  }, [allDomains, overflowingItemIndex, selectedDomain]);
+
+  return (
+    <div className="flex h-[34px] w-fit overflow-hidden whitespace-nowrap rounded-lg border border-solid border-gray-200">
+      <AllTeamsItem selected={selectedDomain === undefined} />
```
Comment (mmioana) [inline src/components/v5/shared/TeamFilter/partials/DesktopTeamFilter/DesktopTeamFilter.tsx:97]: Can we extract this into a constant `const isAllTeamsFilterActive = selectedDomain === undefined` and then use it here
```suggestion
      <AllTeamsItem selected={isAllTeamsFilterActive} />
```

File: src/components/v5/shared/TeamFilter/partials/DesktopTeamFilter/DesktopTeamFilter.tsx
Comment (mmioana) [inline src/components/v5/shared/TeamFilter/partials/DesktopTeamFilter/DesktopTeamFilter.tsx:1]: Would it make sense to extract all `useEffect` and `useMemo` into a separate hook?

File: src/components/v5/shared/TeamFilter/partials/LeftButton.tsx [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { CaretLeft } from '@phosphor-icons/react';
+import React from 'react';
+
+export const LeftButton = () => {
+  return (
+    <div className="flex h-full items-center rounded-l-lg border border-gray-200 bg-base-bg px-4 py-2 text-gray-400">
+      <CaretLeft size={14} />
```
Comment (mmioana) [inline src/components/v5/shared/TeamFilter/partials/LeftButton.tsx:7]: `18` also here

File: src/components/v5/shared/TeamFilter/partials/RightButton.tsx [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { CaretRight } from '@phosphor-icons/react';
+import React from 'react';
+
+export const RightButton = () => {
+  return (
+    <div className="flex h-full items-center rounded-r-lg border border-gray-200 bg-base-bg px-4 py-2 text-gray-400">
+      <CaretRight size={14} />
```
Comment (mmioana) [inline src/components/v5/shared/TeamFilter/partials/RightButton.tsx:7]: `18` also here

File: src/components/v5/shared/TeamFilter/partials/TeamsDropdown/TeamsDropdown.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+          ref={setTooltipRef}
+          {...getTooltipProps()}
+          className="flex flex-col gap-2 rounded-b-lg border-b border-l border-r border-gray-200 bg-base-white px-4 py-2"
+        >
+          {domains.map((domain) => (
+            <button
+              type="button"
+              key={`TeamsDropdown.${domain.nativeId}`}
+              onClick={() => handleClick(domain)}
+              className="w-full text-sm font-medium text-gray-700 hover:font-semibold hover:text-gray-900"
```
Comment (mmioana) [inline src/components/v5/shared/TeamFilter/partials/TeamsDropdown/TeamsDropdown.tsx:59]: We need to left-align the text for all items in the dropdown, so please add `text-start`

File: src/components/v5/shared/TeamFilter/partials/TeamItem.tsx [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      )}
+      onClick={handleClick}
+    >
+      {domain.metadata?.name ?? domain.id}
+    </button>
+  );
+};
+
+TeamItem.displayName = displayName;
+export default TeamItem;
```
Comment (mmioana) [inline src/components/v5/shared/TeamFilter/partials/TeamItem.tsx:51]: We could make use of `useScrollIntoView` for mobile

```suggestion
import clsx from 'clsx';
import React, { type FC, useEffect } from 'react';
import { useSearchParams } from 'react-router-dom';
import { useMobile } from '~hooks';
import { useScrollIntoView } from '~hooks/useScrollIntoView';

import { TEAM_SEARCH_PARAM } from '~routes/index.ts';
import { type Domain } from '~types/graphql.ts';
import { getTeamColor } from '~utils/teams.ts';

const displayName = 'v5.shared.TeamFilter.partials.TeamItem';

interface TeamItemProps {
  selected: boolean;
  domain: Domain;
}

const TeamItem: FC<TeamItemProps> = ({ domain, selected }) => {
  const [searchParams, setSearchParams] = useSearchParams();
  const { ref: teamItemRef, scroll } = useScrollIntoView<HTMLButtonElement>()
  const isMobile = useMobile();

  const teamColor = getTeamColor(domain.metadata?.color);

  useEffect(() => {
    if (selected && isMobile) {
      scroll({
        block: 'end',
        inline: 'center',
        behavior: 'smooth'
      })
    }
  }, [selected, isMobile])

  const handleClick = () => {
    if (selected) {
      searchParams.delete(TEAM_SEARCH_PARAM);
      setSearchParams(searchParams);
    } else {
      searchParams.set(TEAM_SEARCH_PARAM, domain.nativeId.toString());
      setSearchParams(searchParams);
    }
  };

  return (
    <button
      type="button"
      className={clsx(
        'w-full bg-base-white px-4 py-2 text-3',
        selected ? teamColor : null,
        {
          'border-r border-solid border-gray-200 font-medium text-gray-700':
            !selected,
          'border-0 font-semibold text-base-white': selected,
        },
      )}
      onClick={handleClick}
      ref={teamItemRef}
    >
      {domain.metadata?.name ?? domain.id}
    </button>
  );
};

TeamItem.displayName = displayName;
export default TeamItem;
```

**Review by rdig — APPROVED** [review]
This was suuuuuper complex to write, and it's going to take even more sanity away from you until you actually get to merge it.

I've tested kinda all the interactions you listed plus some more, and found it mostly acceptable _(minus all the styling issues that Sam pointed to)_

I'll first list all the testing I've done, then I'll highlight the "weird" parts of the UI I think that we should be aware of, even if we might not necessarily want to change/fix.


https://github.com/user-attachments/assets/fe92a64a-f27f-41d2-8032-e7846740c4cd

https://github.com/user-attachments/assets/d102cb23-6e72-4c74-add8-1830f1df5912

https://github.com/user-attachments/assets/3c03fffc-dc89-4d8d-b2b9-c68be8370291

https://github.com/user-attachments/assets/aaf8727a-1996-40e8-b161-dec8eb970865

If you keep clicking the ">" icon, to scroll through the domains, you can get into a pattern where it will run out of them, show the "+" and you keep clicking bringing up the _Create New Domain_ action

https://github.com/user-attachments/assets/353a913f-235f-415d-bcff-2ac649e2edf8

A selected domain which appears at the end of the list will get hidden by the UI, when the viewport is resized.

I actually don't think this is a big issue, as you can scroll to it using the ">" icon, but is more of a thing we should be aware of

https://github.com/user-attachments/assets/7716916e-0f51-4a71-8b9e-636cd4ec682e

**Review by mmioana — CHANGES_REQUESTED** [review]
Awesome hardcode work on this one @bassgeta 🥇 

Tested on desktop
Screenshot 2024-09-16 at 13 46 42
Screenshot 2024-09-16 at 13 46 53
Screenshot 2024-09-16 at 13 46 58


https://github.com/user-attachments/assets/bf426362-43e4-4b4e-b063-f994ffe87606

Screenshot 2024-09-16 at 13 45 24

And on mobile

Screenshot 2024-09-16 at 13 47 43
Screenshot 2024-09-16 at 13 47 52
Screenshot 2024-09-16 at 13 47 57

And new colony for the empty state
Screenshot 2024-09-16 at 13 48 37


I left some minor refactoring comments, would highly appreciate if you can tackle them 🙏

**Review by mmioana — APPROVED** [review]
Thank you for doing the changes @bassgeta! Stellar work on this one 🌟🥇 


https://github.com/user-attachments/assets/ac6fcff8-8863-4246-b1be-6b4c3238d291


https://github.com/user-attachments/assets/036f09c1-d15c-4910-ac65-fa7d84d1399d


https://github.com/user-attachments/assets/375d8c97-330e-4753-afe0-00fa5eb50464


And the selected filter also scrolls into view 🤩 

https://github.com/user-attachments/assets/7f8e9c38-11fc-4e95-b8ae-44b46074feb3

**Review by iamsamgibbs — APPROVED** [review]
Everything looks great, nice job, lets get it merged!

Screenshot 2024-09-18 at 07 56 25

**Pull request merged**